shell.nix

```diff
@@ -0,0 +1,58 @@
+
+# This shell.nix provides a simple development environment for the btnl_client python package.
+#
+# First, run `nix-shell` to get into the development environment:
+#
+# $ nix-shell
+#
+# This puts you in a shell with a `python` binary on the `PATH`.  The `python` binary is
+# setup to be able to find all required python dependencies (like the
+# `requests` and `websockets` packages).
+#
+# From within this Nix shell, you can run the `bntl_client` module like the
+# following to play around with it:
+#
+# $ python -m btnl_client
+#
+# Here is an example of running one of the available subcommands.  This gets
+# the specifications for a Product:
+#
+# $ python -m btnl_client get-product-spec 3
+# ProductSpreadSpec(type='spread', product_id=3, product_name='Bitcoin US Dollar Calendar Spread', ...
```

Comment:
I don't actually know python at all, so I'm not sure whether or not these are the best ways to run the tool.

But it did seem to work for me at least, so I imagine it is better than nothing.

Reply:
That's basically how I ran it, I think this is perfect

Reply:
hm, even on this pr, after I run `nix-shell` I get the following error
```
bsherman@brad-ThinkPadP16 [09:40:09 AM] [~/code/python-btnl-client] [nix-shell]
-> % nix-shell
[nix-shell] bsherman@brad-ThinkPadP16 [09:40:13 AM] [~/code/python-btnl-client] [nix-shell]
-> % python -m btnl_client get-product-spec 3
Traceback (most recent call last):
  File "/nix/store/2xij13by2fj9mzi16bagf3yb7rim4ify-python3-3.9.20/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/nix/store/2xij13by2fj9mzi16bagf3yb7rim4ify-python3-3.9.20/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/bsherman/code/python-btnl-client/btnl_client/__main__.py", line 5, in <module>
    import btnl_client.websocket as ws
  File "/home/bsherman/code/python-btnl-client/btnl_client/websocket.py", line 7, in <module>
    import websockets
ModuleNotFoundError: No module named 'websockets'
```

Reply:
hm, even on this pr, after I run `nix-shell` I get the following error
```
bsherman@brad-ThinkPadP16 [09:40:09 AM] [~/code/python-btnl-client] [nix-shell]
-> % nix-shell
[nix-shell] bsherman@brad-ThinkPadP16 [09:40:13 AM] [~/code/python-btnl-client] [nix-shell]
-> % python -m btnl_client get-product-spec 3
Traceback (most recent call last):
  File "/nix/store/2xij13by2fj9mzi16bagf3yb7rim4ify-python3-3.9.20/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/nix/store/2xij13by2fj9mzi16bagf3yb7rim4ify-python3-3.9.20/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/bsherman/code/python-btnl-client/btnl_client/__main__.py", line 5, in <module>
    import btnl_client.websocket as ws
  File "/home/bsherman/code/python-btnl-client/btnl_client/websocket.py", line 7, in <module>
    import websockets
ModuleNotFoundError: No module named 'websockets'
```

Reply:
i'm trying to repro on my machine. if i can reproduce the problem you're having, i think changing `mkDerivation` to `mkShell` might fix the problem.

Reply:
Sorry, @bradsherman. I can't reproduce your issue. It "just works™" for me (frustratingly).

I still would suggest you try changing `mkDerivation` to `mkShell`, and see if that fixes your issue.

Reply:
In Nixpkgs, `python*.withPackages` works by creating a wrapped Python binary.  The Python binary is wrapped with some environment variables setup so that it can easily find the libraries you're wrapping it with.

For instance, imagine I make the current change on top of the latest commit here, 12e627048d7af:

```diff
diff --git a/shell.nix b/shell.nix
index 10abd9c..484a367 100644
--- a/shell.nix
+++ b/shell.nix
@@ -47,12 +47,4 @@ let
         ]
       );
 in
-pkgs.stdenv.mkDerivation {
-  name = "python-btnl-client";
-  buildInputs = [
-    python-with-my-packages
-
-    # Other non-python dependencies:
-    pkgs.which
-  ];
-}
+python-with-my-packages
```

And then I go and build this:

```console
$ nix-build ./shell.nix 
/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env
```

If you inspect this derivation output, you can see it contains a couple Python binaries, and a python directory / site-packages thing that contains links to all the libraries we've specified:

```console
$ ls /nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/bin/python
/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/bin/python
$ ls -d -l /nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/lib/python3.9/site-packages/websockets*
lrwxrwxrwx 1 illabout illabout 108 Jan  1  1970 /nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/lib/python3.9/site-packages/websockets -> /nix/store/0kp1gc8czsva7k86jcpib63kg93axhhx-python3.9-websockets-10.4/lib/python3.9/site-packages/websockets
lrwxrwxrwx 1 illabout illabout 123 Jan  1  1970 /nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/lib/python3.9/site-packages/websockets-10.4.dist-info -> /nix/store/0kp1gc8czsva7k86jcpib63kg93axhhx-python3.9-websockets-10.4/lib/python3.9/site-packages/websockets-10.4.dist-info
```

This Python binary appears to just be a wrapper that sets up some environment variables before exec'ing the actual Python binary:

```console
$ strings /nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/bin/python
...
NIX_PYTHONPREFIX=/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env
NIX_PYTHONEXECUTABLE=/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/bin/python3.9
NIX_PYTHONPATH=/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/lib/python3.9/site-packages
PYTHONNOUSERSITE=true
/nix/store/kkbvwghqjfbrh7q1wg59vnyn7rixclk4-python3-3.9.17/bin/python
# ------------------------------------------------------------------------------------
# The C-code for this binary wrapper has been generated using the following command:
makeCWrapper '/nix/store/kkbvwghqjfbrh7q1wg59vnyn7rixclk4-python3-3.9.17/bin/python' \
    --set 'NIX_PYTHONPREFIX' '/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env' \
    --set 'NIX_PYTHONEXECUTABLE' '/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/bin/python3.9' \
    --set 'NIX_PYTHONPATH' '/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/lib/python3.9/site-packages' \
    --set 'PYTHONNOUSERSITE' 'true'
# (Use `nix-shell -p makeBinaryWrapper` to get access to makeCWrapper in your shell)
# ------------------------------------------------------------------------------------
...
```

On my system, I can directly run this binary:

```console
$ /nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/bin/python -m btnl_client get-product-spec 3
ProductSpreadSpec(type='spread', product_id=3, product_name='Bitcoin US Dollar Calendar Spread', ...
```

If I were to try to directly run the underlying non-wrapped `python` binary, I'd expect it to fail because it couldn't find either the `requests` or `websockets` library:

```console
$ /nix/store/kkbvwghqjfbrh7q1wg59vnyn7rixclk4-python3-3.9.17/bin/python -m btnl_client get-product-spec 3
Traceback (most recent call last):
  File "/nix/store/kkbvwghqjfbrh7q1wg59vnyn7rixclk4-python3-3.9.17/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/nix/store/kkbvwghqjfbrh7q1wg59vnyn7rixclk4-python3-3.9.17/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/illabout/git/python-btnl-client/btnl_client/__main__.py", line 2, in <module>
    from btnl_client.product import BitnomialHttpClient, AuthBitnomialHttpClient
  File "/home/illabout/git/python-btnl-client/btnl_client/product.py", line 7, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
```

I'd also expect to be able to manually do the wrapping and have it work:

```console
```

Reply:
> ```
> bsherman@brad-ThinkPadP16 [09:40:09 AM] [~/code/python-btnl-client] [nix-shell]
> -> % nix-shell
> [nix-shell] bsherman@brad-ThinkPadP16 [09:40:13 AM] [~/code/python-btnl-client] [nix-shell]
> -> % python -m btnl_client get-product-spec 3
> Traceback (most recent call last):
>   File "/nix/store/2xij13by2fj9mzi16bagf3yb7rim4ify-python3-3.9.20/lib/python3.9/runpy.py", line 197, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/nix/store/2xij13by2fj9mzi16bagf3yb7rim4ify-python3-3.9.20/lib/python3.9/runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "/home/bsherman/code/python-btnl-client/btnl_client/__main__.py", line 5, in <module>
>     import btnl_client.websocket as ws
>   File "/home/bsherman/code/python-btnl-client/btnl_client/websocket.py", line 7, in <module>
>     import websockets
> ModuleNotFoundError: No module named 'websockets'
> ```

Oh, now that I take a better look at this, it appears that in this error message, the path here (`/nix/store/2xij13by2fj9mzi16bagf3yb7rim4ify-python3-3.9.20`) is version 3.9.20, while my Python from above is 3.9.17.

I'm guessing this is just the wrong Python binary?

Reply:
ah thank you! That actually helped me figure it out! Check out the following commands
```
-> % which python
/home/bsherman/.nix-profile/bin/python
-> % nix-shell
-> % which python
/home/bsherman/.nix-profile/bin/python
-> % where python
/home/bsherman/.nix-profile/bin/python
/home/bsherman/.nix-profile/bin/python
/home/bsherman/.nix-profile/bin/python
/nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/bin/python
/home/bsherman/.nix-profile/bin/python
/home/bsherman/.nix-profile/bin/python
/home/bsherman/.nix-profile/bin/python
/home/bsherman/.nix-profile/bin/python
/home/bsherman/.nix-profile/bin/python
/home/bsherman/.nix-profile/bin/python
```
then as you might have guessed, if I look at $PATH, /home/bsherman/.nix-profile/bin is ahead of the nix store path even when I am in the nix shell. If I explicitly use the python from the shell it works:
```
-> % /nix/store/p2rpvh0w99dp7j0q7m7axpryvmllqcdn-python3-3.9.17-env/bin/python -m btnl_client get-product-spec 3
ProductSpreadSpec(type='spread', product_id=3, product_name='Bitcoin US Dollar Calendar Spread', max_order_quantity=100, min_block_size=5, price_band_variation=50, price_limit_percentage=0.5, pr
ice_increment=1, first_trading_day='2021-06-09', final_settle_time='2021-12-29T15:30:00Z', daily_open_time='08:30:00', daily_settle_time='14:30:00', symbol='BUSZ21-BUSH22', cqg_symbol='BBUSS1Z21
', product_status='expired', base_symbol='BUS', legs=[{'product_id': 1, 'weight': -1}, {'product_id': 2, 'weight': 1}])
```

I'll chalk this up to an incorrect configuration on my machine